pgx/src/datum/array.rs

```diff
@@ -285,6 +277,91 @@ impl<'a, T: FromDatum> Array<'a, T> {
     }
 }
 
+pub struct VariadicArray<'a, T: FromDatum>(Array<'a, T>);
+
+impl<'a, T: FromDatum + serde::Serialize> serde::Serialize for VariadicArray<'a, T> {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        serializer.collect_seq(self.0.iter())
+    }
+}
+
+impl<'a, T: FromDatum> VariadicArray<'a, T> {
+    /// Create an [`Array`](crate::datum::Array) over an array of [`pg_sys::Datum`](pg_sys::Datum) values and a corresponding array
+    /// of "is_null" indicators
+    ///
+    /// `T` can be [`pg_sys::Datum`](pg_sys::Datum) if the elements are not all of the same type
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe as it can't validate the provided pointer are valid or that
+    ///
+    pub unsafe fn over(
+        elements: *mut pg_sys::Datum,
+        nulls: *mut bool,
+        nelems: usize,
+    ) -> VariadicArray<'a, T> {
+        Self(Array::<T> {
+            _ptr: NonNull::new(std::ptr::null_mut()),
+            raw: None,
+            elem_layout: None,
+            nelems,
+            elem_slice: slice::from_raw_parts(elements, nelems).into(),
+            null_slice: slice::from_raw_parts(nulls, nelems).into(),
+            _marker: PhantomData,
+        })
+    }
+
+    pub fn into_array_type(self) -> *const pg_sys::ArrayType {
+        self.0.into_array_type()
+    }
+
+    // # Panics
+    //
+    // Panics if it detects the slightest misalignment between types,
+    // or if a valid slice contains nulls, which may be uninit data.
+    #[deprecated(
+        since = "0.5.0",
+        note = "this function cannot be safe and is not generically sound\n\
+        even `unsafe fn as_slice(&self) -> &[T]` is not sound for all `&[T]`\n\
+        if you are sure your usage is sound, consider RawArray"
+    )]
+    #[allow(deprecated)]
+    pub fn as_slice(&self) -> &[T] {
+        self.0.as_slice()
+    }
```

Comment:
Having to introduce this only to instantly deprecate it displeases me but at least one test depends on it and VariadicArray needs to reimpl Array's full API in order for this to work, so...